File: java/NextPermutation.java
/*
 * Filename: NextPermutation.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Fri 20 Mar 2015 05:59:58 PM EDT
 */

public class Solution {
	public static void nextPermutation(int[] num) {
		if (num.length < 2) {
			return;
		}

		int i = num.length - 1;
		while (i > 0 && num[i-1] >= num[i]) {
			i--;
		}

		int j;
		int temp;

		if (i > 0) {
			j = num.length - 1;
			while (j >= i && num[j] <= num[i-1]) {
				j--;
			}

			temp = num[j];
			num[j] = num[i-1];
			num[i-1] = temp;
		}

		j = num.length - 1;

		while (j > i) {
			temp = num[j];
			num[j] = num[i];
			num[i] = temp;
			j--;
			i++;
		}

		return;
	}
}

File: cpp/next_permutation.cc
/*
 * Filename: next_permutation.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 24 Jul 2015 02:42:05 PM EDT
 */

#include <iostream>
#include <vector>

using namespace std;

class Solution {
	public:
		void nextPermutation(vector<int>& nums) {
			vector<int>::size_type i = nums.size() - 2;
			while (i >= 0 && nums[i] >= nums[i+1]) {
				i--;
			}

			if (i+1) {
				vector<int>::size_type j = nums.size() - 1;
				while (j > i && nums[j] <= nums[i]) {
					j--;
				}

				int tmp = nums[j];
				nums[j] = nums[i];
				nums[i] = tmp;

				i++;
				j = nums.size() - 1;
				while (j > i) {
					tmp = nums[j];
					nums[j] = nums[i];
					nums[i] = tmp;
					i++;
					j--;
				}
			} else {
				for (vector<int>::size_type j = 0; j < nums.size() / 2; j++) {
					int tmp = nums[j];
					nums[j] = nums[nums.size() - j - 1];
					nums[nums.size() - j - 1] = tmp;
				}
			}
		}
};
